### PR TITLE
feat: idleLiquidityModule-v2

### PR DESCRIPTION
**Pull request by gummy789j**

### CODE REVIEW

File: contracts/actions/ClosePositionOneShot.sol [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+                    liquidity: liquidity,
+                    amount0Min: 0,
+                    amount1Min: 0,
+                    deadline: block.timestamp
+                });
+            nonfungiblePositionManager.decreaseLiquidity(decreaseliquidityparams);
+        }
+
+        ///@dev collect
+        INonfungiblePositionManager.CollectParams memory secondCollectparams = INonfungiblePositionManager
```
Comment (kyle8615) [inline contracts/actions/ClosePositionOneShot.sol]: rename for the secondCollectparams

File: contracts/recipes/RefundGasExpenseRecipes.sol [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+import "./BaseRecipes.sol";
+
+///@notice RefundGasExpenseRecipes allows user to refund gas expense from PositionManager
+contract RefundGasExpenseRecipes is BaseRecipes, IRefundGasExpenseRecipes {
+    constructor(address _registryAddressHolder) BaseRecipes(_registryAddressHolder) {}
+
+    ///@notice refund gas expense from PositionManager
+    ///@param receiver address of the receiver
+    ///@param amount amount of gas expense to refund
+    function refundGasExpense(address payable receiver, uint256 amount) external override whenNotPaused {
```
Comment (kyle8615) [inline contracts/recipes/RefundGasExpenseRecipes.sol]: is it necessary to assign a receiver? That's a trade-off between security and convenience

File: deploy/000_Config.ts [hunk 3]
```diff
@@ -42,7 +42,7 @@ export const Config: DeployConfig = {
   "137": {
     sleep: 5000,
     gasLimit: "10000000",
-    gasPrice: "100000000000",
+    gasPrice: "160000000000",
```
Comment (kyle8615) [inline deploy/000_Config.ts:45]: Why need to raise the gas price?

File: deploy/016_WithdrawNativeToken.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    autoMine: true,
+    gasLimit: Config[chainId].gasLimit,
+    gasPrice: Config[chainId].gasPrice,
+    nonce: 15,
+  });
+
+  await new Promise((resolve) => setTimeout(resolve, Config[chainId].sleep));
+  console.log(":: Deployed WithdrawNativeToken: ", (await deployments.get("WithdrawNativeToken")).address);
+};
+
```
Comment (kyle8615) [inline deploy/016_WithdrawNativeToken.ts:26]: these additional contracts should deploy individually in production env, Would it help that place it in the deploy script, or need to create another script?

File: contracts/modules/IdleLiquidityModuleV2.sol [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+                IWithdrawNativeToken.WithdrawNativeTokenInput({
+                    amount: params.rebalanceFee,
+                    receiver: params.feeReceiver
+                })
+            );
+        }
+
+        res.amount0CollectedFee = ucfOutput.amount0;
+        res.amount1CollectedFee = ucfOutput.amount1;
+        res.amount0Removed = cposOutput.amount0Collected.sub(ucfOutput.amount0);
```
Comment (kyle8615) [inline contracts/modules/IdleLiquidityModuleV2.sol:204]: amount1Removed is the token1 amount deducting the collected fee after this rebalance, right? The naming isn't straightforward.

File: deploy/000_Config.ts (hunk 3 above)
Comment (gummy789j) [reply deploy/000_Config.ts:45]: gas price has been more expensive recently

File: contracts/recipes/RefundGasExpenseRecipes.sol (hunk 2 above)
Comment (gummy789j) [reply contracts/recipes/RefundGasExpenseRecipes.sol]: you mean just return to msg.sender ?

File: deploy/016_WithdrawNativeToken.ts (hunk 4 above)
Comment (gummy789j) [reply deploy/016_WithdrawNativeToken.ts:26]: this script could be call individually or collectively, it's ok with both.

File: contracts/modules/IdleLiquidityModuleV2.sol (hunk 5 above)
Comment (gummy789j) [reply contracts/modules/IdleLiquidityModuleV2.sol:204]: cposOutput.amount0Collected -> pure fee income + pure liquidity removed (becuz we use close position in one shot)
ucfOutput.amount0 would be the pure fee income by calculating

res.amount0CollectedFee -> pure fee income = ucfOutput.amount0
res.amount0Removed -> pure liquidity removed = cposOutput.amount0Collected - ucfOutput.amount0

File: contracts/recipes/RefundGasExpenseRecipes.sol (hunk 2 above)
Comment (kyle8615) [reply contracts/recipes/RefundGasExpenseRecipes.sol]: yes

Comment (gummy789j) [reply contracts/recipes/RefundGasExpenseRecipes.sol]: but we still need to confirm that msg.sender is contract or not ? For security, they are at the same level between msg.sender & receiver input ?  For convenience, i think receiver would be more flexible and it's not too inconvenience ?

**Review by kyle8615 — APPROVED** [review]
LGTM

**Pull request merged**